### PR TITLE
std.format: Refactor package (step 3): Add assert messages.

### PR DESCRIPTION
**Pull request by berni44**

### CODE REVIEW

File: std/format/package.d [hunk 1]
```diff
@@ -792,21 +792,25 @@ private void formatReflectTest(T)(ref T val, string fmt, string formatted, strin
         {
             alias aa1 = val;
             alias aa2 = val2;
-            assert(aa1 == aa2);
+            assert(aa1 == aa2, "Identical associative arrays differ.");
```
Comment (RazvanN7) [inline std/format/package.d:795]: This error message could be improved: "Associative arrays differ, but should be identical"

File: std/format/package.d [hunk 2]
```diff
@@ -792,21 +792,25 @@ private void formatReflectTest(T)(ref T val, string fmt, string formatted, strin
...
 
-            assert(aa1.length == aa2.length);
+            assert(aa1.length == aa2.length, "Length of identical associative arrays differs.");
```
Comment (RazvanN7) [inline std/format/package.d:797]:
```suggestion
            assert(aa1.length == aa2.length, "Lengths of associative arrays differ, but should be equal .");
```

File: std/format/package.d [hunk 3]
```diff
@@ -792,21 +792,25 @@ private void formatReflectTest(T)(ref T val, string fmt, string formatted, strin
...
 
-            assert(aa1.keys == aa2.keys);
+            assert(aa1.keys == aa2.keys, "Keys of identical associative arrays differ.");
```
Comment (RazvanN7) [inline std/format/package.d:799]:
```suggestion
            assert(aa1.keys == aa2.keys, "Keys of associative arrays differ, but should be identical");
```

File: std/format/package.d [hunk 4]
```diff
@@ -792,21 +792,25 @@ private void formatReflectTest(T)(ref T val, string fmt, string formatted, strin
...
 
-            assert(aa1.values == aa2.values);
-            assert(aa1.values.length == aa2.values.length);
+            assert(aa1.values == aa2.values, "Values of identical associative arrays differ.");
```
Comment (RazvanN7) [inline std/format/package.d:801]:
```suggestion
            assert(aa1.values == aa2.values, "Values associative arrays differ, but should be identical");
```

File: std/format/package.d [hunk 5]
```diff
@@ -792,21 +792,25 @@ private void formatReflectTest(T)(ref T val, string fmt, string formatted, strin
...
+            assert(aa1.values.length == aa2.values.length,
+                "Length of values of identical associative arrays differ.");
```
Comment (RazvanN7) [inline std/format/package.d:803]: ditto

File: std/format/package.d [hunk 6]
```diff
@@ -792,21 +792,25 @@ private void formatReflectTest(T)(ref T val, string fmt, string formatted, strin
...
             foreach (i; 0 .. aa1.values.length)
-                assert(aa1.values[i] == aa2.values[i]);
+                assert(aa1.values[i] == aa2.values[i],
+                    "Element of identical associative arrays differs.");
 
             foreach (i, key; aa1.keys)
-                assert(aa1.values[i] == aa1[key]);
+                assert(aa1.values[i] == aa1[key],
+                    "Value of associative array cannot accessed via values.");
```
Comment (RazvanN7) [inline std/format/package.d:810]:
```suggestion
                    "Value of associative array cannot be accessed via values.");
```

File: std/format/package.d [hunk 7]
```diff
@@ -792,21 +792,25 @@ private void formatReflectTest(T)(ref T val, string fmt, string formatted, strin
...
             foreach (i, key; aa2.keys)
-                assert(aa2.values[i] == aa2[key]);
+                assert(aa2.values[i] == aa2[key],
+                    "Value of associative array cannot accessed via values.");
```
Comment (RazvanN7) [inline std/format/package.d:813]:
```suggestion
                    "Value of associative array cannot be accessed via values.");
```

**Review by RazvanN7 — CHANGES_REQUESTED** [review]
Please apply the changes to the other messages that I have not commented on as well.

**Comment by PetarKirov** [comment]
To be honest, I think that only asserts inside user-facing functions should have messages. Adding assert messages inside unit tests is pointless as only library developers are ever going to see them and building the code with `-checkaction=context` should be more than sufficient to provide enough information. Actually having assert messages may be even worse, as (IIRC) even if the  code is compiled with this switch, the special formatting won't be applied and only the manually written message may be displayed, if the assert fails.

I understand that enabling the corresponding check in DScanner is motivating in itself, but I'd argue that many of the DScanner rules need additional tuning to make more practical and useful in general.

**Comment by PetarKirov** [comment]
That said, I wouldn't want to throw away work, so by all means feel free to continue. Just keep this in mind for future module.
To clarify, in case there's any confusion: I'm not against assert messages in general, just those inside unit tests, where the code is self-explanatory without them as well. Every rule has an exception, and if an assert message can provide insightful information it has its place even in unit tests. But my general observation is that it's rarely the case, unless it is verifying higher-level features and conditions that are not obvious.

**Comment by berni44** [comment]
> That said, I wouldn't want to throw away work, so by all means feel free to continue.

That's fine. :-) (Isn't there even a buddhist proverb saying, it's most fun to throw away what you just achieved?)

**Pull request closed without merging**